### ISSUE TITLE
Reading Parquet file through Arrow table fails inconsistently

### ISSUE THREAD

**Issue by caleboverman** [duckdb/duckdb#8440]
### What happens?

I've got a parquet file on S3 that I'm attempting to open with duckdb. The file first gets loaded first as a pyarrow dataset, queried from duckdb, then fetched as an arrow table and converted to pandas. The last step converting from arrow table to pandas dataframe fails inconsistently, raising:

```ArrowException: Unknown error: Wrapping `�
                                         -4145-4d40-af19-86fc6a9e1b48 failed```

This was originally discovered when I was using ibis with the duckdb backend so I filed ibis-project/ibis#6631. Thanks to @gforsyth we've determined that it can be reproduced without using ibis and while using multiple versions of pyarrow. He also pointed out that the file has a large number of row groups (51) which could be causing an issue with parallel reads.

I do have a parquet file I can share separately to help reproduce the issue.

Stacktrace:
```ArrowException                            Traceback (most recent call last)
Cell In[9], line 1
----> 1 arrow_table.to_pandas()
        arrow_table = pyarrow.Table
field1: string
field2: string
----
field1: [["cc3259aa-89ee-441d-b0ac-e9e570bbd0c2","dcb37e18-b2d9-4294-8dd9-2164298e01a5","dcb37e18-b2d9-4294-8dd9-2164298e01a5","56c5bf05-2ac4-4fd7-90a5-08d07f1f5c9b","2238e751-a3f7-40e5-adcd-d7023cf635b5",...,"4ba9efd3-1761-4f78-a9a4-6e728286a4f7","896ec64c-ff3b-45bd-a156-bf2e8fb9f216","896ec64c-ff3b-45bd-a156-bf2e8fb9f216","4ba9efd3-1761-4f78-a9a4-6e728286a4f7","4ba9efd3-1761-4f78-a9a4-6e728286a4f7"]]
field2: [["301727","547656","547656","646033","857526",...,"144279","110775","110775","144279","144279"]]

File ~/opt/anaconda3/envs/fts/lib/python3.10/site-packages/pyarrow/array.pxi:837, in pyarrow.lib._PandasConvertible.to_pandas()

File ~/opt/anaconda3/envs/fts/lib/python3.10/site-packages/pyarrow/table.pxi:4114, in pyarrow.lib.Table._to_pandas()

File ~/opt/anaconda3/envs/fts/lib/python3.10/site-packages/pyarrow/pandas_compat.py:820, in table_to_blockmanager(options={'date_as_object': True, 'deduplicate_objects': True, 'integer_object_nulls': False, 'pool': None, 'safe': True, 'self_destruct': False, 'split_blocks': False, 'strings_to_categorical': False, 'timestamp_as_object': False, 'use_threads': True, ...}, table=pyarrow.Table
field1: string
field2: string
----.....,"144279","110775","110775","144279","144279"]], categories=None, ignore_metadata=False, types_mapper=None)
    818 _check_data_column_metadata_consistency(all_columns)
    819 columns = _deserialize_column_index(table, all_columns, column_indexes)
--> 820 blocks = _table_to_blocks(options, table, categories, ext_columns_dtypes)
        table = pyarrow.Table
field1: string
field2: string
----
field1: [["cc3259aa-89ee-441d-b0ac-e9e570bbd0c2","dcb37e18-b2d9-4294-8dd9-2164298e01a5","dcb37e18-b2d9-4294-8dd9-2164298e01a5","56c5bf05-2ac4-4fd7-90a5-08d07f1f5c9b","2238e751-a3f7-40e5-adcd-d7023cf635b5",...,"4ba9efd3-1761-4f78-a9a4-6e728286a4f7","896ec64c-ff3b-45bd-a156-bf2e8fb9f216","896ec64c-ff3b-45bd-a156-bf2e8fb9f216","4ba9efd3-1761-4f78-a9a4-6e728286a4f7","4ba9efd3-1761-4f78-a9a4-6e728286a4f7"]]
field2: [["301727","547656","547656","646033","857526",...,"144279","110775","110775","144279","144279"]]
        ext_columns_dtypes = {}
        options = {'pool': None, 'strings_to_categorical': False, 'zero_copy_only': False, 'integer_object_nulls': False, 'date_as_object': True, 'timestamp_as_object': False, 'use_threads': True, 'deduplicate_objects': True, 'safe': True, 'split_blocks': False, 'self_destruct': False}
        categories = None
    822 axes = [columns, index]
    823 return BlockManager(blocks, axes)

File ~/opt/anaconda3/envs/fts/lib/python3.10/site-packages/pyarrow/pandas_compat.py:1168, in _table_to_blocks(options={'date_as_object': True, 'deduplicate_objects': True, 'integer_object_nulls': False, 'pool': None, 'safe': True, 'self_destruct': False, 'split_blocks': False, 'strings_to_categorical': False, 'timestamp_as_object': False, 'use_threads': True, ...}, block_table=pyarrow.Table
field1: string
field2: string
----.....,"144279","110775","110775","144279","144279"]], categories=None, extension_columns={})
   1163 def _table_to_blocks(options, block_table, categories, extension_columns):
   1164     # Part of table_to_blockmanager
   1165
   1166     # Convert an arrow table to Block from the internal pandas API
   1167     columns = block_table.column_names
-> 1168     result = pa.lib.table_to_blocks(options, block_table, categories,
        block_table = pyarrow.Table
field1: string
field2: string
----
field1: [["cc3259aa-89ee-441d-b0ac-e9e570bbd0c2","dcb37e18-b2d9-4294-8dd9-2164298e01a5","dcb37e18-b2d9-4294-8dd9-2164298e01a5","56c5bf05-2ac4-4fd7-90a5-08d07f1f5c9b","2238e751-a3f7-40e5-adcd-d7023cf635b5",...,"4ba9efd3-1761-4f78-a9a4-6e728286a4f7","896ec64c-ff3b-45bd-a156-bf2e8fb9f216","896ec64c-ff3b-45bd-a156-bf2e8fb9f216","4ba9efd3-1761-4f78-a9a4-6e728286a4f7","4ba9efd3-1761-4f78-a9a4-6e728286a4f7"]]
field2: [["301727","547656","547656","646033","857526",...,"144279","110775","110775","144279","144279"]]
        options = {'pool': None, 'strings_to_categorical': False, 'zero_copy_only': False, 'integer_object_nulls': False, 'date_as_object': True, 'timestamp_as_object': False, 'use_threads': True, 'deduplicate_objects': True, 'safe': True, 'split_blocks': False, 'self_destruct': False}
        categories = None
        pa.lib = <module 'pyarrow.lib' from '/Users/coverman/opt/anaconda3/envs/fts/lib/python3.10/site-packages/pyarrow/lib.cpython-310-darwin.so'>
        pa = <module 'pyarrow' from '/Users/coverman/opt/anaconda3/envs/fts/lib/python3.10/site-packages/pyarrow/__init__.py'>
        extension_columns = {}
   1169                                     list(extension_columns.keys()))
   1170     return [_reconstruct_block(item, columns, extension_columns)
   1171             for item in result]

File ~/opt/anaconda3/envs/fts/lib/python3.10/site-packages/pyarrow/table.pxi:2771, in pyarrow.lib.table_to_blocks()

File ~/opt/anaconda3/envs/fts/lib/python3.10/site-packages/pyarrow/error.pxi:138, in pyarrow.lib.check_status()

ArrowException: Unknown error: Wrapping `�
                                         -4145-4d40-af19-86fc6a9e1b48 failed

```

### To Reproduce

```python
import pyarrow.dataset as ds
from pyarrow.fs import S3FileSystem
import duckdb

s3filesystem = S3FileSystem(
    access_key="...",
    secret_key="...",
    session_token="...",
)

d = ds.dataset(
    source="ibis-testing/ibis_test.parquet",
    format="parquet",
    partitioning="hive",
    filesystem=s3filesystem,
)

con = duckdb.connect()
con.register("dataset", d)
cursor = con.execute("SELECT * FROM dataset;")
arrow_table = cursor.fetch_arrow_table()
arrow_table
arrow_table.to_pandas()
```

### OS:

macOS

### DuckDB Version:

0.8.1

### DuckDB Client:

Python

### Full Name:

Caleb Overman

### Affiliation:

FanThreeSixty

### Have you tried this on the latest `master` branch?

I have tested with a master build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by gforsyth** [comment]
Hey @Mause -- the above reproducer usually trips for me in <10 runs, but I've run this Arrow-only version for 50 iterations without issue:

```python
from pyarrow.fs import S3FileSystem
import pyarrow.dataset as ds

s3filesystem = S3FileSystem(
    access_key="...",
    secret_key="...",
    session_token="..."",
)
d = ds.dataset(
    source="ibis-testing/ibis_test.parquet",
    format="parquet",
    partitioning="hive",
    filesystem=s3filesystem,
)

## Use DuckDB
# import duckdb
# con = duckdb.connect()
# con.register("dataset", d)
# cursor = con.execute("SELECT * FROM dataset;")
# arrow_table = cursor.fetch_arrow_table()
# df = arrow_table.to_pandas()

## Use Arrow alone
arrow_table = d.to_table()
df = arrow_table.to_pandas()
```

**Comment by Mause** [comment]
Hmm okay - it might be worth passing along the parquet file then? Chances are it's the encoding of the text in the file that's interacting oddly. That's assuming that you can reproduce whilst loading the file from your local disk

**Comment by caleboverman** [comment]
@Mause is there a way I can send you the file without posting publicly? Nothing identifying in it but I'd rather be careful.

I hadn't seen the issue locally previously but had also not spent much time trying. However, with a few tries I was able to trigger the issue locally just now.

**Comment by Mause** [comment]
Feel free to email it to elliana@duckdblabs.com, or send it via Discord (@mause.me)

**Comment by Mause** [comment]
Interestingly, I get a segfault inside AddStringOrBlob rather than an ArrowException

**Comment by caleboverman** [comment]
@Mause just wanted to check the status on this. Appreciate any progress that's been made and I'm happy to help if there's anything I can do. Thanks!

**Comment by hannes** [comment]
@caleboverman please understand that we cannot provide time lines for free community support

**Comment by hannes** [comment]
@gforsyth I have been digging into this. The issue seems to disappear when converting the dataset to a table or reading a table in the first place, e.g. 

`con.register("dataset", d.to_table())`

works fine for me. Is there some sort of different re-use of buffers between datasets and tables, because that's what this looks like to me. Maybe @pdet can have a look.

**Comment by pdet** [comment]
Hi @caleboverman, thanks for reporting this issue. I have pushed a PR with a fix here: #8879

Since you asked for the data you provided not to be public, I can't properly add a test for it. Please validate that it indeed fixed your issue. Also if you can provide a minimal, fully public, reproducible example, it would be highly appreciated.

**Comment by caleboverman** [comment]
@pdet confirmed the issue is fixed - appreciate the effort!

I'll work to get a reproducible test in place that I'm able to share.
